### PR TITLE
Container scanning: BerkeleyDB support

### PR DESCRIPTION
**Pull request by jssblck**
# Overview

Adds the ability to read the package list from `BerkeleyDB`.

## Acceptance criteria

We can now discover RPM system dependencies using the BerkeleyDB backend for RPM.

## Testing plan

Along with automated testing, I performed the following comparison.

From latest release of FOSSA:
```shell
; fossa container analyze centos:7 --experimental-scanner --output | jq
[ INFO] Running container scanning with fossa experimental-scanner!
[ INFO] Inferred registry source: https://index.docker.io/library/centos:7
[ INFO] Downloaded: eeb6ee3f44bd0b5103bb561b4c16bcb82328cfe5809ab675bb17ab3a16c517c9.json
[ INFO] Gzip extracted & downloaded: 2d473b07cdd5f0912cd6f1a703352c82b512407db6b05b43f2553732b55df3bc.tar
[ INFO] Analyzing exported container tarball: /private/var/folders/q7/3nvvpy0d6js28m8lypw3tcx80000gn/T/fossa-container-registry-tmp-d1977d0a6ed3d3a2/image.tar
[ INFO] Analyzing Base Layer
{
  "image": {
    "layers": [
      {
        "layerId": "sha256:174f5685490326fc0a1c0f5570b8663732189b327007e47ff13d2ca59673db02",
        "srcUnits": []
      }
    ],
    "os": "centos",
    "osRelease": "7"
  }
}
```

From the version in this PR (edited for brevity):
```shell
; cabal run fossa -- container analyze centos:7 --experimental-scanner --output | tail -n +2 | jq
[ INFO] Running container scanning with fossa experimental-scanner!
[ INFO] Inferred registry source: https://index.docker.io/library/centos:7
[ INFO] Downloaded: eeb6ee3f44bd0b5103bb561b4c16bcb82328cfe5809ab675bb17ab3a16c517c9.json
[ INFO] Gzip extracted & downloaded: 2d473b07cdd5f0912cd6f1a703352c82b512407db6b05b43f2553732b55df3bc.tar
[ INFO] Analyzing exported container tarball: /private/var/folders/q7/3nvvpy0d6js28m8lypw3tcx80000gn/T/fossa-container-registry-tmp-ebe460f817c6fbd3/image.tar
[ INFO] Analyzing Base Layer
[ INFO] Analyzing berkeleydb project at var/lib/rpm/
{
  "image": {
    "layers": [
      {
        "layerId": "sha256:174f5685490326fc0a1c0f5570b8663732189b327007e47ff13d2ca59673db02",
        "srcUnits": [
          {
            "AdditionalDependencyData": null,
            "Build": {
              "Artifact": "default",
              "Dependencies": [
                {
                  "imports": [],
                  "locator": "rpm-generic+acl#centos#7$x86_64#2.2.51-15.el7"
                },
                {
                  "imports": [],
                  "locator": "rpm-generic+audit-libs#centos#7$x86_64#2.8.5-4.el7"
                },
                {
                  "imports": [],
                  "locator": "rpm-generic+basesystem#centos#7$noarch#10.0-7.el7.centos"
                }
                {
                  "imports": [],
                  "locator": "rpm-generic+bind-license#centos#7$noarch#32:9.11.4-26.P2.el7"
                }
              ],
              "Imports": [
                "rpm-generic+acl#centos#7$x86_64#2.2.51-15.el7",
                "rpm-generic+audit-libs#centos#7$x86_64#2.8.5-4.el7",
                "rpm-generic+basesystem#centos#7$noarch#10.0-7.el7.centos",
                "rpm-generic+bind-license#centos#7$noarch#32:9.11.4-26.P2.el7"
              ],
              "Succeeded": true
            },
            "GraphBreadth": "complete",
            "Manifest": "var/lib/rpm/",
            "Name": "var/lib/rpm/",
            "OriginPaths": [
              "var/lib/rpm/Packages"
            ],
            "Type": "berkeleydb"
          }
        ]
      }
    ],
    "os": "centos",
    "osRelease": "7"
  }
}
```

## References

[ANE-274](https://fossa.atlassian.net/browse/ANE-274)

Merging as a stack with #1032 and #1031.

## Checklist

It looks like we've left other container scanning stuff out of the changelog for now, so I've not added this PR and don't plan to cut a release for this PR.

- [x] I added tests for this PR's change (or explained in the PR description why tests don't make sense).
- [ ] If this PR introduced a user-visible change, I added documentation into `docs/`.
- [ ] If this change is externally visible, I updated `Changelog.md`. If this PR did not mark a release, I added my changes into an `# Unreleased` section at the top.
- [ ] If I made changes to `.fossa.yml` or `fossa-deps.{json.yml}`, I updated `docs/references/files/*.schema.json`. You may also need to update these if you have added/removed new dependency type (e.g. `pip`) or analysis target type (e.g. `poetry`).

### CODE REVIEW

**Comment by jssblck** [comment]
@spatten the review bot originally chose you, so I've requested review from you; if you want someone else to take it lmk and I can post it in channel.

File: src/Strategy/BerkeleyDB.hs [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+  -- Read the file content from disk and send it to the parser via stdin.
+  -- This is necessary because the parser doesn't have access to the file system being read.
+  fileContent <- context "read file content" $ readContentsBS file
+
+  -- Handle the JSON response.
+  (bdbJsonOutput :: [Text]) <- context "read raw blobs" . execJson' cwd (bdbCommand bin) . decodeUtf8 $ B64.encode fileContent
+  bdbByteOutput <- context "decode base64" . traverse fromEitherShow $ B64.decode <$> fmap encodeUtf8 bdbJsonOutput
+  entries <- context "parse blobs" . traverse fromEitherShow $ readPackageInfo <$> fmap BSL.fromStrict bdbByteOutput
+  context "parse package info" $ traverse parsePkgInfo entries
```
Comment (mksuthar) [inline src/Strategy/BerkeleyDB.hs:176]: We will need to account for epoch value. In backend it's not used to query knowledge api, but I think we persist locators with epoch value for rpm.  

```bash
✘ megh@Mac-mini  ~/test  fossa container analyze centos:7 --output | jq | grep "findutils" -C 20
        "metadata": {
          "architecture": "aarch64",
          "epoch": 1,
          "license": "GPLv3+",
          "name": "findutils",
          "release": "6.el7",
          "size": 1888914,
          "sourceRpm": "findutils-4.5.11-6.el7.src.rpm",
          "vendor": "CentOS",
          "version": "4.5.11"
        },
```

- https://github.com/knqyf263/go-rpmdb/blob/9a3bd2ebb923399db2189c5d8963af27368ba2c8/pkg/rpmtags.go#L16

Comment (jssblck) [reply src/Strategy/BerkeleyDB.hs:180]: Nice catch. Fixed!

**Review by csasarak — APPROVED** [review]
This looks good to me.

File: src/App/Fossa/Container/Sources/DockerArchive.hs [hunk 2]
```diff
@@ -193,6 +194,7 @@ runAnalyzers osInfo filters = do
     single
     [ DiscoverFunc (Apk.discover osInfo)
     , DiscoverFunc (Dpkg.discover osInfo)
+    , DiscoverFunc (BerkeleyDB.discover osInfo)
```
[the remark this answers is not in the document]
Comment (jssblck) [reply src/App/Fossa/Container/Sources/DockerArchive.hs:197]: Hmm. With the others, we're specifying where we got the information, I think. The fact that it happens to also be the name of the tool is I think less relevant. I think that the goal of this _strategy_ is to examine BerkeleyDB; the fact that it happens to always produce RPM results is not really relevant to the strategy itself, if that makes sense.

File: src/Types.hs [hunk 3]
```diff
@@ -106,6 +107,7 @@ data DiscoveredProjectType
 projectTypeToText :: DiscoveredProjectType -> Text
 projectTypeToText = \case
   AlpineDatabaseProjectType -> "apkdb"
+  BerkeleyDBProjectType -> "berkeleydb"
```
[the remark this answers is not in the document]
Comment (jssblck) [reply src/Types.hs:110]: I think we would have this strategy output whatever it finds, yes. If we found BDB being used for something else, we would make the strategy produce multiple dependency types. 

Reasoning: I think this is generally how strategies should work; for example including a `git` dependency in a `cargo` project should cause the `Cargo` strategy to emit `git` dependencies alongside `cargo` dependencies. Same idea.

**Pull request merged**